### PR TITLE
DM-8179: Preserve background jobs and statuses beyond a browser session.

### PR DESCRIPTION
**Pull request by loitly**
https://jira.lsstcorp.org/browse/DM-8179
Also, refactor AppDataCntlr.js

Background jobs are saved based on a userKey cookie which remained the same unless inactive for more than 2 weeks.
To test this, starts one or more downloads.
The downloads should be visible in other firefly instances of the same browser.
It should also remain after you reload the application or quit and come back some time later.
Also, cancel and remove are reflected in all instances as well.

### CODE REVIEW

File: src/firefly/java/edu/caltech/ipac/firefly/server/events/WebsocketConnector.java [hunk 1]
```diff
@@ -114,24 +116,39 @@ public void close() {
      * @param type  action type.  Either "app_data.wsConnAdded" or "app_data.wsConnRemoved"
```
Comment (tgoldina) [inline src/firefly/java/edu/caltech/ipac/firefly/server/events/WebsocketConnector.java]: typo in javadoc: 'event though' -> 'even though'

File: src/firefly/java/edu/caltech/ipac/firefly/server/events/WebsocketConnector.java [hunk 2]
```diff
@@ -114,24 +116,39 @@ public void close() {
...
+                        .forEach(q -> {
+                            List<String> l = connInfo.get(q.getChannel());
+                            if (l == null) {
+                                l = new ArrayList<>();
+                                connInfo.put(q.getChannel(), l);
+                            }
+                            if (!l.contains(q.getConnID())) l.add(q.getConnID());
+                        });
+                FluxAction action = new FluxAction(type, connInfo);
+                ServerEventManager.fireAction(action, new ServerEvent.EventTarget(ServerEvent.Scope.SELF, seq.getConnID(), null, null));
```
Comment (tgoldina) [inline src/firefly/java/edu/caltech/ipac/firefly/server/events/WebsocketConnector.java]: Just curious: So each client on the same channel will be notified of every new or closed connection? Why do we need it?

File: src/firefly/java/edu/caltech/ipac/firefly/data/ServerEvent.java [hunk 3]
```diff
@@ -91,6 +88,10 @@ public String toJsonString() {
         return sb.toString();
     }
 
+    public enum Scope {SELF, CHANNEL, USER, WORLD}
+
+public enum DataType {JSON, BG_STATUS, STRING}
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/firefly/data/ServerEvent.java]: it was not on purpose.  i am guessing intelliJ has something to do with this.

File: src/firefly/java/edu/caltech/ipac/firefly/data/ServerEvent.java [hunk 4]
```diff
@@ -100,11 +101,19 @@ public String toJsonString() {
...
+        private String userKey;
 
         public EventTarget() {}
 
         public EventTarget(Scope scope) {
             this.scope = scope;
+            if (scope == ServerEvent.Scope.CHANNEL) {
+                this.channel = ServerContext.getRequestOwner().getEventChannel();
+            } else if (scope == Scope.USER) {
+                this.userKey =ServerContext.getRequestOwner().getUserKey();
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/firefly/data/ServerEvent.java]: yes.

File: src/firefly/java/edu/caltech/ipac/firefly/server/events/ServerEventManager.java [hunk 5]
```diff
@@ -32,50 +32,57 @@
...
-        else {
-            sev = new ServerEvent(Name.ACTION, scope, ServerEvent.DataType.JSON, actionStr);
-        }
-
+    /**
+     * Send this JSON string action to the client based on the given scope
+     * @param actionStr
+     * @param scope
+     */
+    public static void fireJsonAction(String actionStr, ServerEvent.Scope scope) {
```
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/firefly/server/events/ServerEventManager.java]: This is intended.  The idea is that you should be able to send an action to self(connId), user, channel, or world without having to create a target.

File: src/firefly/java/edu/caltech/ipac/firefly/server/events/WebsocketConnector.java (hunk 2 above)
Comment (loitly) [inline src/firefly/java/edu/caltech/ipac/firefly/server/events/WebsocketConnector.java]: This is done to support JS external viewer and python API(?).  I am guessing this information may be useful in the future as well.

**Comment by loitly** [comment]
Good catch in regard to email sync.  It's fixed in the latest commit.  As for email not sent, it will be fixed in DM-8182.  Thank you for the review.

**Pull request merged**